File: LeetCode_451/main.cpp
/**
    LeetCode_451
    https://leetcode.com/problems/sort-characters-by-frequency/description/
    MatrixDeity, 2017.
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> um;
        for (char c : s)
            ++um[c];
        auto pred = [&] (char a, char b) {
            return um[a] > um[b] || (um[a] == um[b] && a < b);
        };
        sort(s.begin(), s.end(), pred);
        return s;
    }
};

int main() {
    return 0;
}

File: LeetCode_451/main.go
/**
  LeetCode_451
  https://leetcode.com/problems/sort-characters-by-frequency/
  MatrixDeity, 2022.
*/

package leetcode451

import "sort"

func frequencySort(s string) string {
	counter := map[byte]int{}
	res := []byte(s)

	for _, char := range res {
		counter[char]++
	}

	sort.Slice(res, func(i, j int) bool {
		return counter[res[j]] < counter[res[i]] || (counter[res[i]] == counter[res[j]] && res[i] < res[j])
	})

	return string(res)
}
